pytorch_lightning/plugins/training_type/ipu.py

```diff
@@ -0,0 +1,380 @@
+
+        def to_tuple(x):
+            return tuple(x)
+
+        def to_tensor(x):
+            return torch.tensor(x).unsqueeze(0).repeat(self._n_replicate)
+
+        args = apply_to_collection(args, dtype=list, function=to_tuple)
+        args = apply_to_collection(args, dtype=(int, float), function=to_tensor)
+        return args
+
+    def training_step(self, *args, **kwargs):
+        args = self._prepare_input(args)
+        return self.poptorch_models[RunningStage.TRAINING](*args, **kwargs)
+
+    def validation_step(self, *args, **kwargs):
+        args = self._prepare_input(args)
+        return self.poptorch_models[RunningStage.VALIDATING](*args, **kwargs)
+
+    def test_step(self, *args, **kwargs):
+        args = self._prepare_input(args)
+        return self.poptorch_models[RunningStage.TESTING](*args, **kwargs)
+
+    def predict_step(self, *args, **kwargs):
+        args = self._prepare_input(args)
+        return self.poptorch_models[RunningStage.PREDICTING](*args, **kwargs)
+
+    def teardown(self) -> None:
+        for model in self.poptorch_models.values():
+            model.destroy()
+
+    def _compiled(self, model: Any):
+        # Required to ensure we only attach compiled models, as they are compiled lazily.
+        return model._executable is not None
+
+    def _detach_models(self):
+        """
+        Detaches all stage specific models from IPU devices.
+        """
+        for k, model in self.poptorch_models.items():
+            if self._compiled(model) and model.isAttachedToDevice():
+                model.detachFromDevice()
+
+    def _load_model(self, stage: str):
+        """
+        Loads the stage specific accelerator model onto device if compiled and not attached to IPU devices.
+        Args:
+            stage: The stage to load
+        """
+        self._detach_models()
+        model = self.poptorch_models[stage]
+        if self._compiled(model) and not model.isAttachedToDevice():
+            model.attachToDevice()
+
+    def on_train_start(self):
+        self._load_model(RunningStage.TRAINING)
+
+    def on_validation_start(self):
+        self._load_model(RunningStage.VALIDATING)
+
+    def on_test_start(self):
+        self._load_model(RunningStage.TESTING)
+
+    def on_predict_start(self):
+        self._load_model(RunningStage.PREDICTING)
+
+    def on_train_end(self):
+        self._detach_models()
+
+    def on_validation_end(self):
+        self._detach_models()
+
+    def on_test_end(self):
+        self._detach_models()
+
+    def on_predict_end(self):
+        self._detach_models()
+
+    def on_train_batch_start(self, batch: Any, batch_idx: int, dataloader_idx: int) -> None:
+        # Updates optimizer stats if LR scheduler modified the optimizer state
+        optimizer = self.lightning_module.trainer.optimizers[0]
+        self.poptorch_models[RunningStage.TRAINING].setOptimizer(optimizer)
+
+    @property
+    def on_gpu(self) -> bool:
+        return False
```

Comment:
Shouldn't this be False by default in the Training Plugin? 🤔

Reply:
Hmm we extended the parallel plugin, and the default is to use the `root_device` looking at the `device` assuming its a `torch.device`, but in our case we just return `None`